acipher: add dynamic algorithm selection and decryption support

Akshay-Belsare opened the pull request:
- Extend acipher example to support for RSA algorithms:
  - TEE_ALG_RSAES_PKCS1_V1_5
  - TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA1/224/256/384/512
- Users can select algorithm and key size at runtime via: `optee_example_acipher <key_size> <string> <algo>`
- Supported key sizes: 2048, 3072, 4096 bits
- Defaults to TA_ALG_PKCS1_V1_5 if no algorithm is specified
- Enhances the acipher example for flexible testing of multiple RSA modes

etienne-lms commented on acipher/host/main.c:
```diff
@@ -52,6 +66,32 @@ static void get_args(int argc, char *argv[], size_t *key_size, void **inbuf,
+		} else if (strcmp(algo, "TA_ALG_PKCS1_V1_5") == 0) {
+			*algo_num = TA_ALG_PKCS1_V1_5;
+		} else {
+			printf("%s algo is invalid\n", algo);
```
For consistency:
```suggestion
			fprintf(stderr, "%s algo is invalid\n", algo);

etienne-lms commented on acipher/ta/acipher_ta.c:
```diff
@@ -81,10 +110,21 @@ static TEE_Result cmd_enc(struct acipher *state, uint32_t pt,
 	if (res) {
-		EMSG("TEE_AllocateOperation(TEE_MODE_ENCRYPT, %#" PRIx32 ", %" PRId32 "): %#" PRIx32, alg, key_info.keySize, res);
+		EMSG("TEE_AllocateOperation(TEE_MODE_ENCRYPT, %#"
+		     PRIx32 ", %" PRId32 "): %#" PRIx32,
```
The show argument do not match the order used in `TEE_AllocateOperation()`. Not a big issue but one may found that confusing.

etienne-lms commented on acipher/ta/include/acipher_ta.h:
```diff
@@ -22,4 +22,12 @@
+#define TA_ALG_OAEP_MGF1_SHA256		3
+#define TA_ALG_OAEP_MGF1_SHA384		4
+#define TA_ALG_OAEP_MGF1_SHA512		5
+
```
Maybe use GP TEE Internal Core API algorithm IDs? as suggestion by @jenswi-linaro in another P-R?
Nitpicking: could you remove the extra empty line.

amey-raghatate committed:
acipher: Add dynamic algorithm selection and decryption support

Add support for new RSA PKCS OAEP algorithms:
- TEE_ALG_RSAES_PKCS1_V1_5
- TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA1
- TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA224
- TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA256
- TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA384
- TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA512

Also add support to select algorithm at runtime. The user can now run
the `optee_example_acipher <key_size> <string to encrypt> <algo name>`
command, and the specified algorithm will be used during
encryption/decryption.

The user can now invoke:
optee_example_acipher <key_size> <string to encrypt> <algo name>

Supported values for <key_size> are:
- 2048 bits
- 3072 bits
- 4096 bits

Supported values for <algo name> are:
- TA_ALG_PKCS1_V1_5
- TA_ALG_OAEP_MGF1_SHA1
- TA_ALG_OAEP_MGF1_SHA224
- TA_ALG_OAEP_MGF1_SHA256
- TA_ALG_OAEP_MGF1_SHA384
- TA_ALG_OAEP_MGF1_SHA512

If no algorithm is specified, TA_ALG_PKCS1_V1_5 is selected by default.

Based on the input, the corresponding algorithm is selected and used for
encryption/decryption operations. This enhancement improves flexibility
by allowing users to test different RSA modes using a single binary.

Signed-off-by: Amey Avinash Raghatate <ameyavinash.raghatate@amd.com>
Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>
Acked-by: Jerome Forissier <jerome.forissier@linaro.org>

jforissier commented:
LGTM overall but since this is changing the TA interface, I think the UUID should be changed too (it's just an example so we're probably not too concerned about providing some kind of backward compatibility). What do others think?

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

Akshay-Belsare commented:
Thanks @jforissier!
Tags are now applied.
CC: @jenswi-linaro @etienne-lms

jenswi-linaro commented:
> LGTM overall but since this is changing the TA interface, I think the UUID should be changed too (it's just an example so we're probably not too concerned about providing some kind of backward compatibility). What do others think?

I don't mind changing the UUID, but it might be more trouble than it's worth.

jforissier commented:
> > LGTM overall but since this is changing the TA interface, I think the UUID should be changed too (it's just an example so we're probably not too concerned about providing some kind of backward compatibility). What do others think?
> 
> I don't mind changing the UUID, but it might be more trouble than it's worth.

I see what you mean. Downstream scripts with hardcoded UUIDs etc. :wink: fair enough. Ignore my comment.

Akshay-Belsare commented:
@jenswi-linaro @jforissier May we proceed to close this item?
